Add .NET + Kubernetes blog (#2116)
pulumi/docs: content/blog/managing-kubernetes-infrastructure-with-dotnet-and-pulumi/index.md

---
title: "Managing Kubernetes Infrastructure with .NET and Pulumi"
authors: ["luke-hoban"]
tags: ["dotnet","kubernetes"]

> date: "2019-12-04"

Reviewer: Update to actual publish date.

> date: "2019-12-04"

Reviewer 2: ```suggestion
date: "2019-12-10"
```

meta_desc: "Manage Kubernetes with .NET using the Pulumi.Kubernetes resource provider for Pulumi."

> meta_image: "todo.png"

Author: Pending adding a proper meta image.

Reviewer 2: made one - see if this works for you:
https://drive.google.com/file/d/1j-K1Xm7rwUjLK21RckDdzrSfykHfdmuF/view?usp=sharing

> meta_image: "todo.png"

Reviewer 2: ```suggestion
meta_image: "k8s-dotnet.png"
```

---

> Last month, we announced [.NET support for Pulumi](), including support for AWS, Azure, GCP and many other clouds.  One of the biggest questions we heard was about Kubernetes - can I also use Pulumi to manage my Kubernetes infrastructure in C# and F#, like I can already in TypeScript and Python today?  With last week's release of [`Pulumi.Kubernetes`]() on Nuget, you can now also deploy Kubernetes infrastructure from .NET.

Reviewer: Specify links

> Last month, we announced [.NET support for Pulumi](), including support for AWS, Azure, GCP and many other clouds.  One of the biggest questions we heard was about Kubernetes - can I also use Pulumi to manage my Kubernetes infrastructure in C# and F#, like I can already in TypeScript and Python today?  With last week's release of [`Pulumi.Kubernetes`]() on Nuget, you can now also deploy Kubernetes infrastructure from .NET.

Reviewer: Deliberately not mentioning VB?

> Using .NET to build our Kubernetes infrastruture offers several benefits:

Reviewer 2: ```suggestion
Using .NET to build our Kubernetes infrastructure offers several benefits:
```

# Tour of Kubernetes with .NET

We can see a simple example in practice - deploying an Nginx pod into a Kubernetes cluster.

```cs
var pod = new Pod("pod", new PodArgs
{
    Spec = new PodSpecArgs
    {
        Containers =
        {
            new ContainerArgs
            {
                Name = "nginx",
                Image = "nginx",
            },
        },
    },
});
```

We can deploy this with `pulumi up`.

```
$ pulumi up
Previewing update (luke-dev):

     Type                    Name                              Plan       
 +   pulumi:pulumi:Stack     basic_dotnet_kubernetes-luke-dev  create     
 +   └─ kubernetes:core:Pod  pod                               create     
 
Resources:
    + 2 to create

Do you want to perform this update? yes
Updating (luke-dev):

     Type                    Name                              Status      
 +   pulumi:pulumi:Stack     basic_dotnet_kubernetes-luke-dev  created     
 +   └─ kubernetes:core:Pod  pod                               created     
 
Resources:
    + 2 created

Duration: 11s

> Permalink: https://app.pulumi.com/lukehoban/basic_dotnet_kubernetes/luke-dev/updates/81

Reviewer: Nit: Strip the permalink from the output here?

```

> As always, Pulumi programs describe the desired state of our infrasttructure, instead of an imperative process to construct that infrastructure.  This means that if we change our program, Pulumi will compute the minimal delta to apply to our Kubernetes cluster to transition to this new desired state.  This almost feel like using Edit-and-Continue on our deployed Kubernetes resources - modifying our Kubernetes resources in place inside the cluster.

Reviewer 3: ```suggestion
As always, Pulumi programs describe the desired state of our infrastructure, instead of an imperative process to construct that infrastructure.  This means that if we change our program, Pulumi will compute the minimal delta to apply to our Kubernetes cluster to transition to this new desired state.  This almost feels like using Edit-and-Continue on our deployed Kubernetes resources---modifying our Kubernetes resources in place inside the cluster.
```

We can add a label to our Pod:

```patch
 var pod = new Pod("pod", new PodArgs
 {
+    Metadata = new ObjectMetaArgs
+    {
+        Labels = { { "app", "nginx" } },
+    },
     Spec = new PodSpecArgs
     {
         Containers =
         {
             new ContainerArgs
             {
                 Name = "nginx",
                 Image = "nginx",
             },
        },
    },
});
```

> And then see that this just updates the pod in place with this new label.

Reviewer 2: ```suggestion
And then see that this updates the pod in place with this new label.
```

```
$ pulumi up
Previewing update (luke-dev):

     Type                    Name                              Plan       Info
     pulumi:pulumi:Stack     basic_dotnet_kubernetes-luke-dev             
 ~   └─ kubernetes:core:Pod  pod                               update     [diff: ~metadata]
 
Resources:
    ~ 1 to update
    1 unchanged

Do you want to perform this update? details
  pulumi:pulumi:Stack: (same)
    [urn=urn:pulumi:luke-dev::basic_dotnet_kubernetes::pulumi:pulumi:Stack::basic_dotnet_kubernetes-luke-dev]
    ~ kubernetes:core/v1:Pod: (update)
        [id=default/pod-wyjq31t3]
        [urn=urn:pulumi:luke-dev::basic_dotnet_kubernetes::kubernetes:core/v1:Pod::pod]

>         [provider=urn:pulumi:luke-dev::basic_dotnet_kubernetes::pulumi:providers:kubernetes::default_1_4_0_alpha_1574471668_g18eab106_dirty::127c0c24-8b12-4010-a544-5f5390f91e4e]

Reviewer 3: Looks like you aren't using `1.3.2-preview` ;)

      ~ metadata: {
          ~ labels: {
              + app: "nginx"
            }
        }

Do you want to perform this update? yes
Updating (luke-dev):

     Type                    Name                              Status      Info
     pulumi:pulumi:Stack     basic_dotnet_kubernetes-luke-dev              
 ~   └─ kubernetes:core:Pod  pod                               updated     [diff: ~metadata]
 
Resources:
    ~ 1 updated
    1 unchanged

Duration: 3s

Permalink: https://app.pulumi.com/lukehoban/basic_dotnet_kubernetes/luke-dev/updates/82
```

> The real benefits of .NET come when we extract common code into a reusable component.  We can do that to create a new component like a `ServiceDeployment` which includes both a Kubernetes `Service` and `Deployment` using opinionated defaults.  With this, we can describe entire Kubernetes applications (100s of lines of YAML), in a short and semantically meaningful snippet of C#:

Reviewer 2: ```suggestion
The real benefits of .NET come when we extract common code into a reusable component.  We can do that to create a new component like a `ServiceDeployment`, which includes both a Kubernetes `Service` and `Deployment` using opinionated defaults.  With this, we can describe entire Kubernetes applications (100s of lines of YAML), in a short and semantically meaningful snippet of C#:
```

    Ports = { 6379 },
});

var redisReplica = new ServiceDeployment("redis-slave", new ServiceDeploymentArgs
{
    Image = "gcr.io/google_samples/gb-redisslave:v1",
    Ports = { 6379 },
});

var frontend = new ServiceDeployment("frontend", new ServiceDeploymentArgs
{
    Replicas = 3,
    Image = "gcr.io/google-samples/gb-frontend:v4",
    Ports = { 80 },
    AllocateIPAddress = true,
    ServiceType = isMiniKube ? "ClusterIP" : "LoadBalancer",
});

return new Dictionary<string, object?>{
    { "frontendIp", frontend.IpAddress },
};
```

You can check out the implementation of this `ServiceDeployment` component in the [Guestbook example](https://github.com/pulumi/examples/tree/master/kubernetes-cs-guestbook/components) in the Pulumi Examples repo on GitHub.

## Cloud + Kubernetes with .NET

Because Pulumi lets you work with both Kubernetes and cloud (AWS, Azure, GCP, and more), you can also create and manage the infrastructure that builds both a managed Kubernetes cluster as well as deploying applications and services into the cluster. Using a single familiar programming model, we have access to all these various cloud technologies at our fingertips.

> For example, we can deploy a managed Kubernetes cluster on GKE, and then deploy a Pod into it:

Reviewer 4: This isn't GKE - it's DO

> ```csharp

Reviewer 5: Some of these snippets use `cs` and this one is using `csharp`. Are these equivalent?

Author: Yes - but I’ll make it consistent in markdown.

Reviewer: `csharp` is preferred as that's what the lang chooser tabs use (on other pages). So, ideally we'd use `csharp` consistently throughout the site. But I could also tweak the chooser to support `cs` as well.

> var cluster = new KubernetesCluster("do-cluser", new KubernetesClusterArgs {

Reviewer 2: ```suggestion
var cluster = new KubernetesCluster("do-cluster", new KubernetesClusterArgs {
```

        NodeCount = nodeCount
    },
});

var k8sProvider = new Pulumi.Kubernetes.Provider("do-k8s", new Pulumi.Kubernetes.ProviderArgs {
    KubeConfig = cluster.KubeConfigs.Apply(array => array[0].RawConfig)
});

var app = new Pulumi.Kubernetes.Apps.V1.Deployment("do-app-dep", new DeploymentArgs {
    Spec = new DeploymentSpecArgs {
        Selector = new LabelSelectorArgs {
            MatchLabels = new InputMap<string>() {
                {"app", "app-nginx"},
            },
        },
        Replicas = appReplicaCount,
        Template = new PodTemplateSpecArgs {
            Metadata = new ObjectMetaArgs() {
                Labels = new InputMap<string>() {
                    {"app", "app-nginx"},
                },
            },
            Spec = new PodSpecArgs() {
                Containers = new ContainerArgs() {
                    Name = "nginx",
                    Image = "nginx",
                }
            }
        }
    },
}, new CustomResourceOptions {
    Provider = k8sProvider
});

var appService = new Pulumi.Kubernetes.Core.V1.Service("do-app-svc", new ServiceArgs {
    Spec = new ServiceSpecArgs() {
        Type = "LoadBalancer",

>         Selector = app.Spec.Apply(spec => spec.Template.Metadata.Labels),

Reviewer 5: Is the auto-lifting support only for TS? i.e. you can't write this as

`app.Spec.Template.Metadata.Labels`?

Author: Lifting is only in JS and Python.  

I think @cyrusnajmabadi and @mikhailshilkov had an idea about how to provide a similar experience in C# (using extension methods), but it requires a much more heavyweight code generation solution. Are we tracking that idea in an issue anywhere?

Reviewer 3: Yes, it's in #3470, which I'm in (early) progress of tearing apart with extended descriptions

        Ports = new ServicePortArgs {
            Port = 80,
        }
    }
}, new CustomResourceOptions {
    Provider = k8sProvider
});

var ingressIp = appService.Status.Apply(status => status.LoadBalancer.Ingress[0].Ip);

if (!string.IsNullOrWhiteSpace(domainName)) {
    var domain = new Domain("do-domain", new DomainArgs() {
        Name = domainName,
        IpAddress = ingressIp,
    });

    var cnameRecord = new DnsRecord("do-domain-cname", new DnsRecordArgs {
        Domain = domain.Name,
        Type = "CNAME",
        Name = "www",
        Value = "@",
    });
}
```

> Note that this example deploys resources first into DigitalOcean (a Kubernetes `Cluster`), then into Kubernetes itself (`Deployment` and `Service` via a `Pulumi.Kubernetes.Provider` configured to connect to the Kuberentes cluster in Digital Ocean), then optionally also more resources in DigitalOcean dependent on the Kubernetes `Service` (`Domain` and `DnsRecord`).  This is a fairly complex infrastructure deployment orchestration, all in a few dozen lines of declarative and strongly typed C# code. 

Reviewer 3: Not sure how many .NET devs have heard of DO... Is it a lot more code to make it AKS/EKS?

Author: Unfortunately yes - due to the mess around reconstructing `kubeconfig`.  We could possibly just hide that behind a function that we don't include in the example though.  I'll see if we can do that.  

I'll admit though - I kind of liked that this example is (a) quite simple (b) shows some of the diversity of cloud support incidentally and (c) shows cloud resources dependent on Kubernetes resources.

Author: After looking again - going to stick with Digital Ocean for this post.

> Note that this example deploys resources first into DigitalOcean (a Kubernetes `Cluster`), then into Kubernetes itself (`Deployment` and `Service` via a `Pulumi.Kubernetes.Provider` configured to connect to the Kuberentes cluster in Digital Ocean), then optionally also more resources in DigitalOcean dependent on the Kubernetes `Service` (`Domain` and `DnsRecord`).  This is a fairly complex infrastructure deployment orchestration, all in a few dozen lines of declarative and strongly typed C# code. 

Reviewer 2: ```suggestion
Note that this example deploys resources first into DigitalOcean (a Kubernetes `Cluster`), then into Kubernetes itself (`Deployment` and `Service` via a `Pulumi.Kubernetes.Provider` configured to connect to the Kubernetes cluster in Digital Ocean), then optionally also more resources in DigitalOcean dependent on the Kubernetes `Service` (`Domain` and `DnsRecord`).  This is a fairly complex infrastructure deployment orchestration, all in a few dozen lines of declarative and strongly typed C# code. 
```

# Conclusion

> Kubernetes support is one of several significant new additions to the Pulumi .NET support, and [many more improvements](https://github.com/pulumi/pulumi/issues/3470) are in progress over the coming weeks. [Get started](https://www.pulumi.com/docs/get-started/kubernetes/) with Kubernetes and .NET today, and let us know what you think! 

Author: Ideally we would add Kubernetes + .NET support to the Getting Started workflow before publishing this.

Author: This is in https://github.com/pulumi/docs/pull/2129, which itself depends on https://github.com/pulumi/templates/pull/77.

Author: These are both merged now.

> Kubernetes support is one of several significant new additions to the Pulumi .NET support, and [many more improvements](https://github.com/pulumi/pulumi/issues/3470) are in progress over the coming weeks. [Get started](https://www.pulumi.com/docs/get-started/kubernetes/) with Kubernetes and .NET today, and let us know what you think! 

Reviewer: ```suggestion
Kubernetes support is one of several significant new additions to the Pulumi .NET support, and [many more improvements](https://github.com/pulumi/pulumi/issues/3470) are in progress over the coming weeks. [Get started]({{< relref "/docs/get-started/kubernetes" >}}) with Kubernetes and .NET today, and let us know what you think! 
```
